proof.go

```diff
@@ -131,56 +131,82 @@ type ProofStep struct {
 	neighbor     ProofStepNeighbor
 }
 
+func prependTag(tag byte, encodedValue []byte) []byte {
+	// d8 is the indicator for a 1-byte tag value
+	return append([]byte{0xd8, tag}, encodedValue...)
+}
+
 func (s *ProofStep) MarshalCBOR() ([]byte, error) {
+	var valueToTag any
+	var tagValue byte
+	var err error
+
 	switch s.stepType {
 	case ProofStepTypeBranch:
+		tagValue = 121 // 0x79
 		tmpNeighbors := []byte{}
 		for _, neighbor := range s.neighbors {
-			tmpNeighbors = append(tmpNeighbors, neighbor.Bytes()...)
+			if neighbor == (Hash{}) {
+				tmpNeighbors = append(tmpNeighbors, NullHash[:]...)
+			} else {
+				tmpNeighbors = append(tmpNeighbors, neighbor.Bytes()...)
+			}
 		}
-		tmpData := cbor.NewConstructor(
```

Comment:
Why replace the use of `cbor.Constructor` here? You're adding magic numbers (`121`, `122`, and `123` for the CBOR tag types) and don't seem to be gaining anything